src/main/java/frc/robot/subsystems/SwerveSubsystem.java

```diff
@@ -152,13 +153,14 @@ private SwerveSubsystem() {
     updateFeedforward = new UpdateSimpleFeedforward((ff) -> updateModuleFeedforward(ff), swerveTable, Config.Swerve.driveKS, Config.Swerve.driveKV, Config.Swerve.driveKA);
   }
 
-  public void drive(
-      ChassisSpeeds speeds, boolean fieldRelative, boolean isOpenLoop) {
+  public void drive(ChassisSpeeds speeds, boolean fieldRelative, boolean isOpenLoop, boolean flipForAlliance) {
+    Rotation2d heading = flipForAlliance ? rotateForAlliance(getHeading()) : getHeading();
+
     SwerveModuleState[] swerveModuleStates =
     Config.Swerve.swerveKinematics.toSwerveModuleStates(
       // ChassisSpeeds.discretize(
         fieldRelative ? 
-            ChassisSpeeds.fromFieldRelativeSpeeds(speeds, rotateForAlliance(getHeading())) :
+            ChassisSpeeds.fromFieldRelativeSpeeds(speeds, heading) :
```

Comment:
why do we need to take account for blue/red alliance here?
Auto starting position will set it correctly.

Reply:
On the blue alliance, forwards is defined as a heading of 0. On the red alliance, forwards is defined as a heading of 180 degrees. But the drivers still want forwards to be away from them. So on the red alliance, the driver controls need to rotate the heading by 180 degrees to account for this.

Reply:
FieldRelativeSpeeds: forward means away from our alliance wall. 
Agreed, for teleop, we do need to adjust by ourselves. In auto, pathPlanner takes care.
"On the blue alliance, forwards is defined as a heading of 0." This is set by PathPlanner?

Reply:
We draw paths for the blue alliance, pathplanner mirrors them to the red alliance. PathPlanner resets our odometry and has defined "The origin of the field coordinate system will remain on the blue side." This means that, even when on the red alliance, driving towards a heading of 0 will always be driving away from the blue alliance station wall.

But the drivers want the robot to drive the robot according to their perspective, which on the red side is 180 degrees opposite of what odometry is set to by pathplanner. So to fix this, a bunch of headings have to be rotated by 180 degrees when on the red side.

Reply:
yeah, I found the words in pathPlanner, i.e. "The origin of the field coordinate system will remain on the blue side."
Thanks for explanation.